### PR TITLE
feat: handle post status through `is_public` meta

### PR DESCRIPTION
**Pull request by miguelpeixe**
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

This PR introduces "controlled statuses". `publish` or `private` statuses changes are now capable of publishing a newsletter. The `is_public` meta will determine which is the final status of a sent newsletter.

This approach introduces the use of `private` status to control newsletter visibility and enables simplification of how we hook into global queries to avoid non-public newsletters from being displayed.

### How to test the changes in this Pull Request:

1. Check out this branch, make sure you have an ESP configure and send a newsletter that will not be public
2. Confirm the newsletter is sent and the post is `private`
3. Edit the sent newsletter, toggle the "Make newsletter page public?" option, update and confirm the post is now "publish"
4. Toggle off the option, save and confirm it's back to private
5. Send the newsletter to the trash and restore
6. Confirm the restored newsletter is set to private
7. Send a new newsletter with "Make newsletter page public?" toggled on and confirm it's sent with status set to "publish"

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

**Comment by miguelpeixe** [comment]
Thanks, @dkoo!

14dac8146afb5a1d6d11bde6f6abefed678a3eea is definitely a good call, changing the visibility option from that UI will take no effect so it's best to keep it away from the publisher. Is this label tweak common? It feels unusual to have a checkbox label changing its text according to the selection.

I wonder if we should do 3e07ed92b91a164bb21a0a2888f6413976fbdda6 on a different PR, along with the migration strategy. The way this is implemented does not require the query changes to take effect immediately. Newsletters marked as `publish` will continue to use the query until we resolve the migration and created/updated newsletters will be set to private and handled by core.

**Comment by dkoo** [comment]
> Is this label tweak common? It feels unusual to have a checkbox label changing its text according to the selection.

I'm not sure—we do it on some of our other repos, but I suppose @thomasguillot can weigh in on that?

> I wonder if we should do https://github.com/Automattic/newspack-newsletters/commit/3e07ed92b91a164bb21a0a2888f6413976fbdda6 on a different PR, along with the migration strategy. The way this is implemented does not require the query changes to take effect immediately. Newsletters marked as publish will continue to use the query until we resolve the migration and created/updated newsletters will be set to private and handled by core.

That's a great point. I've reverted that change and the changes to the unit tests so we can handle on a separate PR.

**Comment by miguelpeixe** [comment]
> we do it on some of our other repos

No objection to this :)

Requesting @adekbadek review, since @dkoo and I contributed here. It's also good to have another set of eyes on this. It's not a lot of code, but have a big impact on how newsletters are handled.

**Comment by thomasguillot** [comment]
> > Is this label tweak common? It feels unusual to have a checkbox label changing its text according to the selection.
> 
> I'm not sure—we do it on some of our other repos, but I suppose @thomasguillot can weigh in on that?

I think this is fine 👍

**Review by adekbadek — APPROVED** [review]

File: includes/service-providers/class-newspack-newsletters-service-provider.php [hunk 1]
```diff
@@ -103,14 +110,19 @@ public function pre_post_update( $post_id, $data ) {
...
-		if ( 'publish' === $old_status && 'publish' !== $new_status && $sent ) {
+		// Prevent status change from the controlled status if newsletter has been sent.
+		if ( ! in_array( $new_status, self::$controlled_statuses, true ) && $old_status !== $new_status && $sent ) {
 			$error = new WP_Error( 'newspack_newsletters_error', __( 'You cannot change a sent newsletter status.', 'newspack-newsletters' ) );
 			wp_die( $error ); // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
 		}
 
-		// Send if changing from any status to publish.
-		if ( 'publish' === $new_status && 'publish' !== $old_status ) {
+		// Send if changing from any status to controlled statuses - 'publish' or 'private'.
```
[the remark this answers is not in the document]
Comment (miguelpeixe) [reply includes/service-providers/class-newspack-newsletters-service-provider.php:119]: The status change being a trigger to send newsletters is subject to change. The fact that they are controlled by the plugin has a bigger meaning and impact on how the system works.

**Pull request merged**